core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java

```diff
@@ -363,6 +344,10 @@ public RetryOnReplicaException(StreamInput in) throws IOException {
     }
 
     private final class AsyncReplicaAction extends AbstractRunnable {
+        /**
+         * The number of operations remaining before we can reply. See javadoc for {@link #operationComplete()} more.
+         */
+        private final AtomicInteger operationsUntilReply = new AtomicInteger(2);
```

Comment:
what's wrong with synchronized and volatile? I mean they can't be set concurrently that makes assertions much simpler?